kong/tools/dns_resolver.lua

```diff
@@ -0,0 +1,81 @@
+local Object = require "classic"
+local stringy = require "stringy"
+local cache = require "kong.tools.database_cache"
+local resolver = require "resty.dns.resolver"
+
+local DEFAULT_TIMEOUT = 2000
+local DEFAULT_ATTEMPTS = 5
+
+local DnsResolver = Object:extend()
+
+function DnsResolver:new(host, port, timeout, attempts)
+  self.host = host
+  self.port = port
+  self.timeout = timeout or DEFAULT_TIMEOUT
+  self.attempts = attempts or DEFAULT_ATTEMPTS
+end
+
+function DnsResolver:query(address, type)
+  -- Init the resolver
+  local r, err = resolver:new{
+    nameservers = {{self.host, self.port}},
+    retrans = self.attempts,
+    timeout = self.timeout
+  }
+  if not r then
+    return nil, "Startup error: "..err
+  end
+
+  -- Make query
+  local answers, err = r:query(address, {qtype = r["TYPE_"..type]})
+  if not answers then
+    return nil, "failed to query the DNS server: ", err
+  end
+
+  if answers.errcode then
+    return nil, "server returned error code: ", answers.errcode, ": ", answers.errstr
+  end
+
+  return answers
+end
+
+function DnsResolver:resolve(address, port, enable_srv)
+  -- Retrieve from cache
+  local cache_key = cache.dns_key(address)
+  local result = cache.get(cache_key)
+  if result then
+    return {host = result.host, port = result.port}
+  end
+
+  -- Query for A record
+  local answers, err = self:query(address, "A")
+  if err then
+    return nil, err
+  end
+  if #answers <= 0 then
+    return nil, "could not find any record for "..address
+  end
+
+  local a_answer = answers[1]
+  local final_address = a_answer.address
+
+  -- Query for SRV record
+  local final_port = port
+  if enable_srv then
+    local answers, err = self:query(address, "SRV")
+    if not err and #answers > 0 then -- Ignore the error because some DNS servers don't support SRV
+      local srv_answer = answers[1]
+      if srv_answer.port > 0 then
+        final_port = srv_answer.port
+      end
+    end
+  end
+
+  if a_answer.ttl > 0 then
+    cache.set(cache_key, {host = final_address, port = final_port}, a_answer.ttl)
```

Comment:
my results;
````
MacBook-Pro:kong thijs$ dig srvtest.thijsschreijer.nl SRV

; <<>> DiG 9.8.3-P1 <<>> srvtest.thijsschreijer.nl SRV
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33090
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;srvtest.thijsschreijer.nl.	IN	SRV

;; ANSWER SECTION:
srvtest.thijsschreijer.nl. 599	IN	SRV	20 10 8000 10.0.0.44.
srvtest.thijsschreijer.nl. 599	IN	SRV	10 5 8001 www.thijsschreijer.nl.
srvtest.thijsschreijer.nl. 599	IN	SRV	10 5 8000 10.0.0.43.

;; Query time: 31 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Tue May 17 23:47:21 2016
;; MSG SIZE  rcvd: 142

MacBook-Pro:kong thijs$
````
The entries with the highest priorities should be used, in this case (priority 20, 10, 10), it means the 2 records with prio 10 (lower = higher priority). An then do a weighted round robin on that. Weights being 5 and 5.

So probably something like;

- if the name has no `_` do an A lookup
- if it fails, or there is an `_` do an SRV lookup
- if SRV returns a hostname, resolve that one

Because the default ttl = 0 with Consul, its going to be a hot path, so we shouldn't do A and SRV unless absolutely necessary.